model/quic-socket-tx-scheduler.h

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2020 SIGNET Lab, Department of Information Engineering, University of Padova
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation;
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ * Authors: Federico Chiariotti <chiariotti.federico@gmail.com>
+ *          Michele Polese <michele.polese@gmail.com>
+ *          Umberto Paro <umberto.paro@me.com>
+ *
+ */
+
+#ifndef QUICSOCKETTXSCHEDULER_H
+#define QUICSOCKETTXSCHEDULER_H
+
+#include "quic-socket-tx-buffer.h"
+
+namespace ns3
+{
+
+class QuicSocketTxItem;
+
+/**
+ * \ingroup quic
+ *
+ * \brief Tx socket buffer for QUIC
+ */
+class QuicSocketTxScheduler : public Object
+{
+public:
+  /**
+   * \brief Get the type ID.
+   * \return the object TypeId
+   */
+  static TypeId GetTypeId (void);
+
+  /**
+   * Print the scheduler buffer information to a string
+   *
+   * \param os the std::ostream object
+   */
+  virtual void Print (std::ostream &os) const = 0;
+
+  //friend std::ostream & operator<< (std::ostream & os, QuicSocketTxBuffer const & quicTxBuf);
+
+  /**
+   * Add a tx item to the scheduling list
+   *
+   * \param item a smart pointer to a transmission item
+   */
+  virtual void Add (Ptr<QuicSocketTxItem> item, bool retx) = 0;
+
+  /**
+   * \brief Get the next scheduled packet with a specified size
+   *
+   * \param numBytes number of bytes of the QuicSocketTxItem requested
+   * \return the item that contains the right packet
+   */
+  virtual Ptr<QuicSocketTxItem> GetNewSegment (uint32_t numBytes) = 0;
+
+  /**
+   * Returns the total number of bytes in the application buffer
+   *
+   * \return the total number of bytes in the application buffer
+   */
+  virtual uint32_t AppSize (void) const = 0;
+
+};
+
+/**
+ * \brief The FIFO implementation
+ *
+ * This class is a simple FIFO implementation of the socket scheduler, which treats all streams the same
+ */
+class QuicSocketTxFifoScheduler : public QuicSocketTxScheduler
+{
+public:
+  /**
+   * \brief Get the type ID.
+   * \return the object TypeId
+   */
+  static TypeId GetTypeId (void);
+
+  QuicSocketTxFifoScheduler ();
+
+  QuicSocketTxFifoScheduler (const QuicSocketTxFifoScheduler &other);
+
+  ~QuicSocketTxFifoScheduler (void);
+
+  /**
+   * Print the scheduler buffer information to a string
+   *
+   * \param os the std::ostream object
+   */
+  void Print (std::ostream &os) const;
+
+  void Add (Ptr<QuicSocketTxItem> item, bool retx);
```

Comment:
override

Reply:
This does not override anything, QuicSocketTxScheduler doesn't inherit from anyone